aten/src/ATen/native_parse.py

```diff
@@ -186,6 +186,73 @@ def parse_arguments(args, func_variants, declaration, func_return):
+    # types and names of keyword arguments: "ScalarType dtype", "Layout layout" and "Device device"
+    # Each, if set, must have default arguments set to long or float, strided and "cpu" respectively.
+    # They must appear in this order and in this order only in order for us to be able to process them.
+    # In the future we will get rid of this specific processing as downstream consumers start relying
+    # less on the content of Declarations.yaml. If you want to support more than this you'll
+    # potentially have to extend the JIT.
+
+    currently_supported_tensor_options_arguments = [
+        # dtype is specified as an int64_t of at::ScalarType
+        {'name': 'dtype', 'type': ['ScalarType', 'ScalarType?'], 'is_nullable': False, 'annotation': None},
+        # layout is specified as an int64_t of at::Layout
+        {'name': 'layout', 'type': ['Layout', 'Layout?'], 'is_nullable': False, 'annotation': None},
+        # device is specified as an IntArrayRef of { at::Device::Type, device_id }
+        {'name': 'device', 'type': ['Device', 'Device?'], 'is_nullable': False, 'annotation': None},
+    ]
+    tensor_options_defaults = [['c10::nullopt'], ['c10::nullopt'], ['c10::nullopt']]
+
+    def compare_tensor_option(argument, tensor_option_argument):
+        matches = True
+        matches = matches and argument['name'] == tensor_option_argument['name']
+        matches = matches and argument['type'] in tensor_option_argument['type']
+        return matches
+
+    def is_tensor_option(argument, currently_supported_tensor_options_arguments):
+        is_in_it = False
+        for to_argument in currently_supported_tensor_options_arguments:
+            is_in_it = is_in_it or compare_tensor_option(argument, to_argument)
+        return is_in_it
+
+    new_arguments = []
+    idx = 0
+    while idx < len(arguments):
+        argument = arguments[idx]
+        if is_tensor_option(argument, currently_supported_tensor_options_arguments) and \
+                len(arguments) - idx >= 3:
+            tensor_options_representation = []
+            for i in range(3):
+                argument = arguments[idx]
+                if not is_tensor_option(argument, currently_supported_tensor_options_arguments):
+                    break
+                tensor_options_representation.append(argument)
+                idx += 1
+            if len(tensor_options_representation) == 3:
+                merged_argument = {'type': 'TensorOptions', 'name': 'options',
+                                   'is_nullable': False, 'annotation': None}
+                for t_idx, t_arg in enumerate(tensor_options_representation):
+                    assert compare_tensor_option(t_arg, currently_supported_tensor_options_arguments[t_idx])
+                    if 'default' in t_arg:
+                        assert t_arg['default'] in tensor_options_defaults[t_idx]
+                        if 'default' not in merged_argument:
```

Comment:
doesn't this allow a bunch of unsupported combinations?  E.g. some of these parameters can have defaults, some not, some can be kwargs, some not, allows both individual options and TensorOptions parameters, etc.